runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java

```diff
@@ -760,6 +776,24 @@ private void addDisplayData(Step step, String stepName, HasDisplayData hasDispla
       List<Map<String, Object>> list = MAPPER.convertValue(displayData, List.class);
       addList(getProperties(), PropertyNames.DISPLAY_DATA, list);
     }
+
+    private void addResourceHints(Step step, String stepName, ResourceHints hints) {
+      Map<String, Object> urlEncodedHints = new HashMap<>();
+      PercentCodec codec = new PercentCodec();
+      for (Entry<String, ResourceHint> entry : hints.hints().entrySet()) {
+        try {
```

Comment:
You could just use https://docs.oracle.com/javase/7/docs/api/java/net/URLEncoder.html#encode(java.lang.String,%20java.lang.String)

Reply:
Looks like this encodes space to +, which would be inconsistent with Python behavior and require some pre or post processing.

Reply:
As long as the consumer unquotes `+` back to space (which is standard) we should be fine. `+` is always quoted to `%2b` to avoid this ambiguity.

Reply:
Another nice thing about Apache Commons codec is that it understands bytearray, URLEncoder will require another string conversion. I think it's fine to keep as is, but if you feel strongly we should change to URLEnconder I'm happy to do that.

Reply:
I don't have a strong opinion here, was mostly thinking about this due to the dependency issues above. (I'm assuming apache commons is not a new dependency, and we don't have to worry about shading, vendoring, and/or versioning issues.)

Reply:
It doesn't seem to be new for Beam, https://github.com/apache/beam/blob/8b7f0b994926981298eb8195b2af24c8f12cebee/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L500, but I had to declare it a compile-time dep for dataflow runner module. It is small (400kb jar) and unlikely cause versioning conflicts.
One thought is that if we ever get to have binary-valued hints, encoding  bytes directly will be easier. I notice though PercentCodec does not by default encode '+' as '%2b', but I can configure that to be safe.